pkg/products/codeready/reconciler.go

```diff
@@ -628,3 +645,123 @@ func getKeycloakClientSpec(cheURL string) keycloak.KeycloakClientSpec {
+		Namespace: namespace,
+		Name:      ruleName,
+	}
+
+	if err := serverClient.Get(ctx, selector, pr); err != nil {
+		return errorUtil.Wrapf(err, "exception calling DeleteRDSAvailabilityAlert: %s", ruleName)
+	}
+
+	// call delete on that object
+	if err := serverClient.Delete(ctx, pr); err != nil {
+		return errorUtil.Wrapf(err, "exception calling DeleteRDSAvailabilityAlert: %s", ruleName)
+	}
+	logrus.Infof(fmt.Sprintf("PrometheusRule: %s reconciled successfully.", pr.Name))
+
+	return nil
+}
+
+// CreateElastiCacheAvailabilityAlert Call this when we create the ElastiCache, to create a
+// PrometheusRule alert to watch for the availability of the ElastiCache instance
+func (r *Reconciler) CreateElastiCacheAvailabilityAlert(ctx context.Context, cr *v1alpha1.Redis, serverClient k8sclient.Client,
+	alertName string, alertNamespace string, alertResourceID string, alertProductName string,
+) error {
+	ruleName := fmt.Sprintf("availability-rule-%s", alertResourceID)
+	alertExp := intstr.FromString(
+		fmt.Sprintf("absent(cro_aws_elasticache_available{exported_namespace='%s',resourceID='%s',productName='%s'} == 1)",
+			cr.Namespace, alertResourceID, alertProductName),
+	)
+	alertDescription := fmt.Sprintf("The product: %s, ElastiCache instance: %s, in namespace: %s is unavailable", alertProductName, alertResourceID, alertNamespace)
+	labels := map[string]string{
+		"severity":    "critical",
+		"productName": cr.Labels["productName"],
+	}
+
+	// CreatePrometheusRule(ruleName string, namespace string, alertRuleName string,
+	//	description string, alertExp intstr.IntOrString, labels map[string]string)
+	pr, err := croResources.CreatePrometheusRule(ruleName, cr.Namespace, alertName, alertDescription, alertExp, labels)
+	if err != nil {
+		return err
+	}
+
+	// Unless it already exists, call the kubernetes api and create this PrometheusRule
+	// Replace this with CreateOrUpdate if we can figure it out
+	err = serverClient.Create(ctx, pr)
+	if err != nil {
+		if !kerrors.IsAlreadyExists(err) {
+			return errorUtil.Wrap(err, fmt.Sprintf("exception calling Create prometheusrule: %s", ruleName))
+		}
+	}
+	logrus.Infof(fmt.Sprintf("PrometheusRule: %s reconciled successfully.", pr.Name))
+	return nil
+}
+
+// DeleteElastiCacheAvailabilityAlert We call this when we delete an ElastiCache instance,
+// it removes the prometheusrule alert which watches for the availability of the instance.
+func (r *Reconciler) DeleteElastiCacheAvailabilityAlert(ctx context.Context, serverClient k8sclient.Client, namespace string, instanceID string) error {
+	// query the kubernetes api to find the object we're looking for
+	ruleName := fmt.Sprintf("availability-rule-%s", instanceID)
+
+	pr := &prometheusv1.PrometheusRule{}
+	selector := client.ObjectKey{
+		Namespace: namespace,
+		Name:      ruleName,
+	}
+
+	if err := serverClient.Get(ctx, selector, pr); err != nil {
+		return errorUtil.Wrapf(err, "exception calling DeleteRDSAvailabilityAlert: %s", ruleName)
```

Comment:
I think if we do a get on the prometheus rule and it doesn't exist (for example, it might have been deleted earlier), we shouldn't need to return an error

Reply:
I think you're right, but should we print a log statement maybe ?